File: crates/homie-remote/tests/prefs_sync.rs
use std::fs;

use homie_remote::{
    HostEntry, PrefsSyncToolReport, mkdir_argv, prefs_sync_specs, present_sync_items, rsync_argv,
    rsync_failure_message,
};
use tempfile::TempDir;

#[test]
fn prefs_sync_uses_fixed_include_list_and_never_credentials() {
    let home = TempDir::new().expect("home");
    write(home.path(), ".claude/CLAUDE.md");
    write(home.path(), ".claude/settings.json");
    write_dir(home.path(), ".claude/commands");
    write_dir(home.path(), ".claude/skills");
    write(home.path(), ".claude/.credentials.json");
    write_dir(home.path(), ".claude/projects");
    write_dir(home.path(), ".claude/todos");
    write(home.path(), ".codex/config.toml");
    write(home.path(), ".codex/auth.json");

    let specs = prefs_sync_specs(home.path());
    let claude = specs.iter().find(|spec| spec.name == "claude").unwrap();
    let codex = specs.iter().find(|spec| spec.name == "codex").unwrap();
    let claude_items = present_sync_items(claude);
    let codex_items = present_sync_items(codex);

    assert_eq!(
        claude_items,
        ["CLAUDE.md", "settings.json", "commands", "skills"]
    );
    assert_eq!(codex_items, ["config.toml"]);

    let host = host();
    let rsyncs = [
        rsync_argv(&host, claude, &claude_items),
        rsync_argv(&host, codex, &codex_items),
    ];
    let flat = rsyncs
        .iter()
        .flatten()
        .cloned()
        .collect::<Vec<_>>()
        .join("\n");
    assert!(!flat.contains("credentials"));
    assert!(!flat.contains("auth.json"));
    assert!(!flat.contains("projects"));
    assert!(!flat.contains("todos"));
    assert!(!flat.contains("--delete"));
    assert!(rsyncs[0].contains(&"-a".to_string()));
    assert!(rsyncs[0].contains(&"--timeout=60".to_string()));
    assert_eq!(
        rsyncs[0].last().map(String::as_str),
        Some("you@forge:.claude/")
    );
    assert_eq!(
        rsyncs[1].last().map(String::as_str),
        Some("you@forge:.codex/")
    );

    let mkdir = mkdir_argv(&host, claude);
    assert_eq!(mkdir.first().map(String::as_str), Some("ssh"));
    assert!(mkdir.iter().any(|arg| arg.contains("ConnectTimeout=10")));
    assert!(mkdir.iter().any(|arg| arg.contains("mkdir -p .claude")));
}

#[test]
fn prefs_sync_empty_config_is_success_without_commands() {
    let home = TempDir::new().expect("home");
    write(home.path(), ".claude/CLAUDE.md");
    let specs = prefs_sync_specs(home.path());
    let codex = specs.iter().find(|spec| spec.name == "codex").unwrap();
    let items = present_sync_items(codex);

    assert!(items.is_empty());
    let report = PrefsSyncToolReport::skipped(codex.name.clone());
    assert!(report.ok);
    assert!(report.synced.is_empty());
    assert_eq!(report.error, None);
}

#[test]
fn maps_missing_remote_rsync_to_clear_error() {
    let message = rsync_failure_message(127, "bash: rsync: command not found", "Forge");
    assert!(message.contains("rsync is not installed on Forge"));

    let generic = rsync_failure_message(12, "protocol error", "Forge");
    assert_eq!(generic, "rsync failed (exit 12): protocol error");
}

fn host() -> HostEntry {
    HostEntry {
        id: "forge".to_string(),
        name: Some("Forge".to_string()),
        ssh: "you@forge".to_string(),
        default_cwd: Some("~/code".to_string()),
        node: None,
    }
}

fn write(root: &std::path::Path, relative: &str) {
    let path = root.join(relative);
    fs::create_dir_all(path.parent().unwrap()).expect("parent");
    fs::write(path, "x").expect("write");
}

fn write_dir(root: &std::path::Path, relative: &str) {
    fs::create_dir_all(root.join(relative)).expect("dir");
}

File: Sources/HomieDaemonKit/PrefsSync.swift
import HomieProtocol
import Foundation

/// Pushes the local user's agent preferences to a remote host so agents there
/// behave like local ones (`host.sync_prefs`). Swift port of
/// `infra/sync-agent-prefs.sh` — users don't have the repo checked out, so the
/// daemon owns the logic.
///
/// The include list is FIXED and additive (`rsync -a`, never `--delete`):
///   claude: CLAUDE.md, settings.json, keybindings.json, commands/, skills/, agents/
///   codex:  config.toml, AGENTS.md, prompts/
/// Excluded on purpose: `.credentials.json` / `auth.json` (each user logs in
/// with their own account on the box), `projects/` (transcripts + memory are
/// per-machine, path-slugged), todos, caches, statsig, shell snapshots.
public enum PrefsSync {
    public struct ToolSpec: Sendable {
        /// Report name ("claude" / "codex").
        public var name: String
        /// Local config directory (e.g. `~/.claude`).
        public var localDir: URL
        /// Remote directory relative to the remote $HOME (e.g. `.claude`).
        public var remoteDir: String
        /// The fixed include list. Anything not named here never syncs.
        public var items: [String]
    }

    public static func toolSpecs(
        home: URL = FileManager.default.homeDirectoryForCurrentUser
    ) -> [ToolSpec] {
        [
            ToolSpec(
                name: "claude",
                localDir: home.appendingPathComponent(".claude", isDirectory: true),
                remoteDir: ".claude",
                items: [
                    "CLAUDE.md", "settings.json", "keybindings.json",
                    "commands", "skills", "agents",
                ]),
            ToolSpec(
                name: "codex",
                localDir: home.appendingPathComponent(".codex", isDirectory: true),
                remoteDir: ".codex",
                items: ["config.toml", "AGENTS.md", "prompts"]),
        ]
    }

    /// Items from the spec's include list that exist locally right now.
    static func presentItems(
        of spec: ToolSpec, fileManager: FileManager = .default
    ) -> [String] {
        spec.items.filter {
            fileManager.fileExists(atPath: spec.localDir.appendingPathComponent($0).path)
        }
    }

    /// `mkdir -p <remoteDir>` over ssh (relative to the remote $HOME).
    static func mkdirArgv(host: HostEntry, spec: ToolSpec) -> [String] {
        ["ssh"] + HostShell.sshOptions
            + [host.ssh, "--", "mkdir -p \(InjectionBuilder.shellQuote(spec.remoteDir))"]
    }

    /// The additive rsync push. `-a` only — no `--delete`, ever — and the ssh
    /// transport carries the same non-interactive options as every other
    /// daemon-initiated remote command.
    static func rsyncArgv(host: HostEntry, spec: ToolSpec, present: [String]) -> [String] {
        let transport = (["ssh"] + HostShell.sshOptions).joined(separator: " ")
        return ["rsync", "-a", "--timeout=60", "-e", transport]
            + present.map { spec.localDir.appendingPathComponent($0).path }
            + ["\(host.ssh):\(spec.remoteDir)/"]
    }

    /// Runs the full sync for every tool. Never throws — each tool reports its
    /// own success/failure so one broken tool doesn't hide the other's result.
    public static func run(
        host: HostEntry,
        runner: ShellRunner,
        home: URL = FileManager.default.homeDirectoryForCurrentUser
    ) async -> HostSyncPrefsResult {
        var reports: [PrefsSyncToolReport] = []
        for spec in toolSpecs(home: home) {
            reports.append(await sync(spec: spec, host: host, runner: runner))
        }
        return HostSyncPrefsResult(tools: reports)
    }

    static func sync(
        spec: ToolSpec, host: HostEntry, runner: ShellRunner
    ) async -> PrefsSyncToolReport {
        let present = presentItems(of: spec)
        guard !present.isEmpty else {
            // Nothing local to push is a success, not an error (matches the
            // reference script's "claude: nothing to sync").
            return PrefsSyncToolReport(tool: spec.name, ok: true, synced: [])
        }
        do {
            let mkdir = try await runner.run(
                mkdirArgv(host: host, spec: spec), nil, .seconds(30))
            guard mkdir.ok else {
                return PrefsSyncToolReport(
                    tool: spec.name, ok: false, synced: [],
                    error: "ssh to \(host.displayName) failed: \(failureDetail(mkdir))")
            }
            let rsync = try await runner.run(
                rsyncArgv(host: host, spec: spec, present: present), nil, .seconds(120))
            guard rsync.ok else {
                return PrefsSyncToolReport(
                    tool: spec.name, ok: false, synced: [],
                    error: rsyncFailureMessage(rsync, host: host))
            }
            return PrefsSyncToolReport(tool: spec.name, ok: true, synced: present)
        } catch {
            return PrefsSyncToolReport(
                tool: spec.name, ok: false, synced: [], error: "\(error)")
        }
    }

    /// Maps an rsync failure to something actionable. The classic trap is a
    /// remote box without rsync installed: the remote shell prints "command
    /// not found" and rsync dies with a protocol error — say so plainly.
    static func rsyncFailureMessage(_ result: ShellResult, host: HostEntry) -> String {
        let stderr = result.stderr.trimmingCharacters(in: .whitespacesAndNewlines)
        if stderr.localizedCaseInsensitiveContains("command not found")
            || stderr.localizedCaseInsensitiveContains("rsync: not found")
            || result.exitCode == 127
        {
            return
                "rsync is not installed on \(host.displayName) — install it there (e.g. apt install rsync) and retry"
        }
        return "rsync failed (exit \(result.exitCode)): \(stderr)"
    }

    private static func failureDetail(_ result: ShellResult) -> String {
        let stderr = result.stderr.trimmingCharacters(in: .whitespacesAndNewlines)
        return stderr.isEmpty ? "exit \(result.exitCode)" : stderr
    }
}
